Feature/android/1315/directions detailed guide (#787)
MapsPeople/docs: src/android/v3/guides/directions/directions-wayfinding-instructions.md

---
title: Get Directions and Show Wayfinding Instructions
eleventyNavigation:
  parent: android-v3-guides-directions
  title: Get Directions and Show Wayfinding Instructions
  key: android-v3-guides-get-directions-and-show-wayfinding
  order: 410
---

In this tutorial we will show how to work with the route model returned from a directions service call. We will also show how you can utilize interaction between the route rendering on the map and textual instructions showed in another view.

To start we will use the complete Getting started example found here: [Java](https://github.com/MapsPeople/MapsIndoors-Getting-Started-Android) or [Kotlin](https://github.com/MapsPeople/MapsIndoors-Getting-Started-Android-Kotlin)

> We will work with changing the implementation of the `NavigationFragment` and the `RouteLegFragment`.

Reviewer: ```suggestion
You will work with changing the implementation of the `NavigationFragment` and the `RouteLegFragment`.
```

```java/0,7
private MPLocation mLocation;
private Route mRoute;

public static NavigationFragment newInstance(Route route, MapsActivity mapsActivity, MPLocation location) {
    final NavigationFragment fragment = new NavigationFragment();
    fragment.mRoute = route;
    fragment.mMapsActivity = mapsActivity;
    fragment.mLocation = location;
    return fragment;
}
```

</mi-tab-panel>
<mi-tab-panel id="kotlin">
<a href="https://github.com/MapsPeople/MapsIndoors-Getting-Started-Android-Kotlin/blob/advanced_directions/app/src/main/java/com/example/mapsindoorsgettingstartedkotlin/NavigationFragment.kt#L147-L155">NavigationFragment.kt</a>

```kotlin/1,7
private var mRoute: Route? = null
private var mLocation: MPLocation? = null

companion object {
    fun newInstance(route: Route?, mapsActivity: MapsActivity?, location: MPLocation?): NavigationFragment {
        val fragment = NavigationFragment()
        fragment.mRoute = route
        fragment.mLocation = location
        fragment.mMapsActivity = mapsActivity
        return fragment
    }
}
```

</mi-tab-panel>
</mi-tabs>

> Now we will replace the old view in the `RouteLegFragment`.

Reviewer: ```suggestion
Next step is replacing the old view in the `RouteLegFragment`.
```

    val routeCollectionAdapter =
        RouteCollectionAdapter(this)
    val mViewPager: ViewPager2 = view.findViewById(R.id.stepViewPager)
    mViewPager.adapter = routeCollectionAdapter
    mViewPager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
        override fun onPageSelected(position: Int) {
            super.onPageSelected(position)
            //When a page is selected call the renderer with the index
            mMapsActivity?.getMpDirectionsRenderer()?.setRouteLegIndex(position)
            //Update the floor on mapcontrol if the floor might have changed for the routing
            mMapsActivity?.getMpDirectionsRenderer()?.currentFloor?.let {floorIndex ->
                mMapsActivity?.getMapControl()?.selectFloor(floorIndex)
            }
        }
    })

    val closeBtn = view.findViewById<ImageView>(R.id.close_btn)

    //Button for closing the bottom sheet. Clears the route through directionsRenderer as well, and changes map padding.
    closeBtn.setOnClickListener {
        mMapsActivity!!.removeFragmentFromBottomSheet(this)
        mMapsActivity!!.getMpDirectionsRenderer()?.clear()
    }
}
```

</mi-tab-panel>
</mi-tabs>

> We will now change the `ViewPager` implementation to contain the revised `RouteLegFragment` for each leg of the route.

Reviewer: ```suggestion
Next step is to change the `ViewPager` implementation to contain the revised `RouteLegFragment` for each leg of the route.
```

```java/0-2,6-8
private String mStep = null;
private int mDuration = 0;
private int mDistance = 0;

public static RouteLegFragment newInstance(String step, int distance, int duration) {
    RouteLegFragment fragment = new RouteLegFragment();
    fragment.mStep = step;
    fragment.mDistance = distance;
    fragment.mDuration = duration;
    return fragment;
}
```

</mi-tab-panel>
<mi-tab-panel id="kotlin">
<a href="https://github.com/MapsPeople/MapsIndoors-Getting-Started-Android-Kotlin/blob/advanced_directions/app/src/main/java/com/example/mapsindoorsgettingstartedkotlin/RouteLegFragment.kt#L46-L53">RouteLegFragment.kt</a>

```kotlin/0-2,6-8
private String mStep = null;
private int mDuration = 0;
private int mDistance = 0;

public static RouteLegFragment newInstance(String step, int distance, int duration) {
    RouteLegFragment fragment = new RouteLegFragment();
    fragment.mStep = step;
    fragment.mDistance = distance;
    fragment.mDuration = duration;
    return fragment;
}
```

</mi-tab-panel>
</mi-tabs>

> We will now update the `onViewCreated` method to use the new views added earlier in the tutorial.

Reviewer: ```suggestion
You must also update the `onViewCreated` method to use the new views added earlier in the tutorial.
```
